Enhance explorer observations with chain time

ffakenz opened the pull request:
<!-- Describe your change here -->

To see the "freshness" of the data:

<img width="1384" alt="image" src="https://github.com/input-output-hk/hydra/assets/14299604/c7e17bda-8d53-46e3-9d28-1461c728d0e3">

To see head details:

<img width="1496" alt="image" src="https://github.com/input-output-hk/hydra/assets/14299604/63c34929-3897-4829-82cc-a05c7714b297">

To see head member commit details:

<img width="1503" alt="image" src="https://github.com/input-output-hk/hydra/assets/14299604/9c38dd51-f789-41dc-b18b-cf20614e80c9">

---

<!-- Consider each and tick it off one way or the other -->
* [X] CHANGELOG updated or not needed
* [X] Documentation updated or not needed
* [X] Haddocks updated or not needed
* [X] No new TODOs introduced or explained herafter

ffakenz committed:
Expose OnChainTx to the observerHandler instead of head observation

This resolves a FIXME: To avoid having to work around NoHeadTx.

ffakenz committed:
Add chain time to head observations

To see the freshness of the data.

ffakenz committed:
Add chain time explorer HeadState

To see the freshness of the date at /heads endpoint.

ffakenz committed:
Update explorer UI to display chain time info per head state

ch1bo commented on hydra-chain-observer/src/Hydra/ChainObserver.hs:
```diff
@@ -162,32 +170,33 @@ chainSyncClient tracer networkId startingPoint observerHandler =
+    OnCloseTx{headId} -> pure $ HeadCloseTx{headId}
+    OnFanoutTx{headId} -> pure $ HeadFanoutTx{headId}
+    OnAbortTx{headId} -> pure $ HeadAbortTx{headId}
+    OnContestTx{headId} -> pure $ HeadContestTx{headId}
```
Must: not use `Maybe` anymore here

ch1bo commented on hydra-chain-observer/src/Hydra/ChainObserver.hs:
```diff
@@ -53,7 +51,17 @@ import Ouroboros.Network.Protocol.ChainSync.Client (
+type ObserverHandler m = [HeadObservationAt] -> m ()
+
+data HeadObservationAt = HeadObservationAt
+  { slotNo :: SlotNo
```
Should: use `ChainPoint` instead as it is unique

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -273,20 +288,27 @@ replaceHeadState newHeadState@HeadState{headId = newHeadStateId} explorerState =
+      HeadObservationAt{slotNo, blockNo, onChainTx = OnContestTx{headId, snapshotNumber}} ->
+        aggregateContestObservation headId slotNo blockNo snapshotNumber explorerState
+      HeadObservationAt{slotNo, blockNo, onChainTx = OnFanoutTx{headId}} ->
+        aggregateFanoutObservation headId slotNo blockNo explorerState
```
Should: DRY

Instead of updating all those `aggregate` functions you could just do handle the `slotNo` and `blockNo` data outside of the `case` for all `HeadObservationAt`

ffakenz replied:
can't because inside the aggregate functions we create unknown head states that require this info.

ch1bo commented on hydra-explorer/test/Hydra/Explorer/ExplorerStateSpec.hs:
```diff
@@ -23,8 +24,8 @@ spec = do
 
   getHeadIds :: ExplorerState -> [HeadId]
-  getHeadIds = fmap headId
+  getHeadIds = fmap ExplorerState.headId
```
This qualified import should not be needed (looking at the imports)

ch1bo reviewed (changes requested):
In principle a fine change, left a requested change though as you have a leftover `Maybe`.

In general, this is not what I meant. It would not improve the "freshness" of the data shown as the chain points displayed in the table would stay whenever that head changed it's state. What I meant, though is to provide information about what was the latest seen block / slot of the whole explorer. So we can see something moving as time passes (even when no heads are updated).

ffakenz committed:
Add slotNo and blockNo to ExplorerState

To see the "freshness" of the data.

ffakenz committed:
Remove the Maybe return in logOnChainTx as not needed anymore

ffakenz committed:
Replace SlotNo by ChainPoint instead as it is unique

ffakenz committed:
Remove qualified import and minor formatting changes

ch1bo commented on hydra-chain-observer/src/Hydra/ChainObserver.hs:
```diff
@@ -162,32 +169,38 @@ chainSyncClient tracer networkId startingPoint observerHandler =
+          let observationsAt =
+                fmap convertObservation observations <&> \case
+                  Just onChainTx -> HeadObservationAt point blockNo (Just onChainTx)
+                  Nothing -> HeadObservationAt point blockNo Nothing
```
Should: model this better

I would say that this is not really a `HeadObservationAt` if it observes `Nothing`, right?

Instead of having a wrapped `Maybe`, you could have a sum type where one is just a `Tick` and another is a `Observation`. You could consider re-using the `ChainEvent` type, which is basically just that!

ffakenz replied:
re-using the `ChainEvent` type its different from what I need.

however I refactored the model to follow the same approach.

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -262,31 +279,74 @@ aggregateFanoutObservation headId explorerState =
+        ExplorerState
+          { heads
+          , point
+          , blockNo
```
Must: display these in the user interface

I don't see where the last seen point/block number is shown in the user interface? I would expect the API yaml to contain it somewhere and the user interface to show it outside of the table as these data points are not related to some specific head.

ffakenz committed:
Refactor chain observation model

To avoid using Maybe type to represent head observations.

ffakenz committed:
Add GET /tick endpoint

To obtain latest point in time observed on chain.

ffakenz committed:
Fix tick chain observations

We were missing to produce tick observations when observed transactions
were no head transactions.

ffakenz committed:
Add web first draft

ffakenz committed:
Minor formatting changes

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -119,12 +139,14 @@ withHydraExplorer cardanoNode mStartChainFrom action =
  where
   getHeads = responseBody <$> (parseRequestThrow "http://127.0.0.1:9090/heads" >>= httpJSON)
 
+  getTick = responseBody <$> (parseRequestThrow "http://127.0.0.1:9090/tick" >>= httpJSON)
```
I guess at some point we want this url to be configurable?

ffakenz committed:
Mova heads data fetcher to provider

ffakenz committed:
Add HeadsDashboard

Enhance web HeadState model.
Refactor HeadsTable into HeadsDataProvider:
- to be able to reuse the /heads fetched data in multiple components.

ffakenz committed:
Add head amd member commits details popups

ffakenz committed:
Display total locked value at every level of the UI

ffakenz committed:
Add scrollbar to HeadDetails

To not overflow when having many participants.

ffakenz committed:
Add utils to calculate totalValueLocked in lovelace

ffakenz committed:
Minor formatting changes

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -26,6 +26,7 @@ type CorsHeaders =
 
+-- REVIEW: maybe rename
 type GetHeadsHeaders :: [Type]
 type GetHeadsHeaders = Header "Accept" String ': CorsHeaders
```
Should: Not only rename, but remove!

The `Accept` header is set by the `'[JSON]` content type handlers and the cors headers we can set on all requests using a middleware. No need to repeat these for each route (at least move them up further in the `API` type).

ffakenz committed:
Fix pointInBlock observed

Also:
- removed unnecessary header type
- sort the heads table based on block number
- set pulling /heads data on by default
- add links to sancho.cexplorer

ffakenz committed:
Minor formatting changes

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -26,41 +26,61 @@ type CorsHeaders =
+  result <- liftIO $ try getTick
+  case result of
+    Right tick -> return tick
     Left (_ :: SomeException) -> throwError err500
```
This is equivalent:

```suggestion
handleGetTick getTick = liftIO getTick
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -73,18 +93,25 @@ logMiddleware tracer app' req sendResponse = do
+readModelGetTick :: TVar IO ExplorerState -> GetTick
+readModelGetTick explorerStateTVar = do
+  ExplorerState{tick} <- readTVarIO explorerStateTVar
+  pure tick
```
Could: DRY this into a single function and provide that to the `httpApp`.

For example, a `getState :: IO ExplorerState` function and then the handler implementations would be:

```
getHeads = getState <&> \ExplorerState{heads} -> heads

getTick = getState <&> \ExplorerState{tick} -> tick
```

ch1bo commented on hydra-explorer/test/Hydra/Explorer/ExplorerStateSpec.hs:
```diff
@@ -16,15 +16,25 @@ spec = do
+                          HeadObservation{} -> False
+                          Tick{} -> True
+                      )
+                 )
```
Should: avoid `suchThat` as it makes inefficient generators.

An alternative implementation could be something like:

```suggestion
  genObservations :: Gen [ChainObservation]
  genObservations = 
    listOf1 $ HeadObservation <$> arbitrary <*> arbitrary <*> arbitrary

```

ch1bo reviewed (approved):
Quite verbose implementation (see comments), but shows the right data.

ch1bo committed:
Move option parsing into hydra-explorer main

This way we can use the Options type as early as possible.

ch1bo committed:
Use wai-cors middleware

This adds the blanket content source policy on all responses handled by
the httpApp.

ch1bo committed:
DRY explorer API handler access on read model

ch1bo committed:
Improve genObservations

Using suchThat may lead to suboptimal generators while creating the
value we want from its constituents is always a better option (if possible).

ch1bo committed:
Fix wai-spec tests of explorer and simplify

Uses the package-specific ResponseMatcher with its MatchHeader and
MatchBody constructors to hook things up with the openapi schema
validation.

ch1bo commented:
@ffakenz I have pushed several improvements on the `hydra-explorer` API and tests. Pinging you for your later reference.
